### DISCUSSION TITLE
partition keys

### DISCUSSION THREAD

**Discussion by lishq** [milvus-io/milvus#29316] [Q&A and General discussion] [answered]
Can a single data record only belong to one unique partition? For example, if the same data is stored in partition A, can it still be stored in partition B?

**Comment by xiaofan-luan** [comment]
It depends on the parition key field.
If partition is same then data should be int he same partition

**Reply by lishq** [reply]
In Milvus, is it possible to assign multiple partition key values to a single data record? For example, can an item such as 'Tofu Pudding' belong to two different categories, like 'Beijing Snack' and 'Shanghai Snack', without duplicating the storage of data? This is to avoid increased memory overhead due to storing and indexing the same data vector multiple times.

**Comment by yhmo** [comment]
@lishq 
Data of different partitions is stored in different storage paths. Milvus storage is column-based. And data is split into segments. Each field of a segment is stored as a column-based file.

Each record is assigned to a partition according to the hash value of its partiiton key. So, if a record is stored in partition_A, it cannot be stored in partition_B.

**Reply by lishq** [reply]
ok. thank you for your patience

**Comment by lishq** [comment]
感谢大家的耐心解答。我还想请教一下，比如多租户的场景。虽然可以通过分区来隔离每个租户的数据，但是假如有一批公共数据，既属于分区A，又属于分区B，但不属于分区C；或者这批公共数据属于所有分区。

我不想在每个分区都冗余一条公共数据的话，这种业务有没有比较好的设计呢。

**Reply by xiaofan-luan** [reply]
可以放在default分区，搜的时候，可以同事搜a分区和default分区

**Reply by lishq** [reply]
那假如有的数据既属于分区b，又属于分区c，但不属于分区a；这种数据我放到了default分区的话，那么搜a分区及公共数据的时候，我怎么把它过滤掉呢。

比如这个例子：豆腐脑是同时属于北京和上海的小吃。为了避免数据冗余，我把这条数据放到了default分区。然后我搜索云南的美食的话，我需要同时搜索“云南”分区和default分区。但是default分区里的“豆腐脑”这个食物不属于[云南]。这个怎么过滤掉呢？或者有没有别的推荐的使用方式
{
  "food":"豆腐脑"，
  "city":["北京","上海"]
},
{
  "food":"米饭"，
  "city":["北京","上海","云南"]
},
{
  "food":"过桥米线"，
  "city":["云南"]
}

**Reply by yhmo** [reply, marked as answer]
分区的主要目的是把数据在物理上隔离，然后可以有效减小搜索的范围： 当我知道我想要的答案在某个特定的分区里，那我搜索时只需搜索一个分区即可，不需要全表搜索，从而大幅减少搜索延迟。
所以，“你以什么方式来减小搜索范围” 就决定了 “你要以什么方式来分区”。分区只有一层，它不像学校那样先按年龄分年级，再按学号在每个年级里分班。你只能用一种标准来分区，你不能既要按A来分，又要按B来分。

但我看你的这些需求描述，貌似并不需要使用向量数据库啊，用传统的关系数据库就可以了。

**Reply by lishq** [reply]
明白了，辛苦耐心解答。我这里是为了理解方便所以举了食物的例子。我的具体需求是要用milvus存储文档，然后向量搜索。就比如A公司有一批文档可以放到A分区、B公司也有一批文档可以放到B分区。但是A公司和B公司可能会有重复或者共享的文档。但是这些文档又不允许C公司看到。所以看看在不冗余存储的情况下，有没有别的方案可以只存一条数据来实现这个需求。因为分享出来的数据可能会比较多的话，重复存储还挺浪费空间的。

**Reply by lishq** [reply]
大佬 再追问一下。milvus可以指定 id in （）后面这个集合最多可以是多少。这个in后面的集合数量太大会不会影响效率

**Reply by yhmo** [reply]
列存储的不太好做成你说的这种“共享某条数据”，列存方便大规模读取，实际上是以空间换时间。像upsert这种操作也是以冗余的方式来实现的，毕竟存储便宜。

"id in []" 我也没验证过能多大，可能上限是RPC传输的最大字节数吧。越大解析起来越慢，越大返回的结果集越多，结果越多传输也越慢。一般还是分批地做比较好，先查 id in [1, 2, 3 ,4 ,5] 再查 id in [6, 7 , 8, 9 ,10]

**Reply by lishq** [reply]
好的 感谢

**Reply by xiaofan-luan** [reply]
一般是通过过滤能力满足你说的需求的

**Reply by lishq** [reply]
是的，过滤比较好。milvus有支持数组类型字段的计划吗？如果能支持的话这个需求过滤起来会简单些。
比如还是我这个例子，我把公共数据存到了default分区后，要查询北京的数据可以通过city字段指定包含“北京”的数据。

{
"food":"豆腐脑"，
"city":["北京","上海"]
},
{
"food":"米饭"，
"city":["北京","上海","云南"]
},
{
"food":"过桥米线"，
"city":["云南"]
}

**Reply by yhmo** [reply]
从2.3.2开始支持Array type： https://milvus.io/docs/array_data_type.md#Array
对Array的过滤可以用 array_contains，看文档里的示例。
